<div align="center">
  <img src="https://github.com/user-attachments/assets/ffe1315d-9e5c-48cd-a712-a3765102a4b5" alt="ProBord" width="450" />
</div>

# ProAct: **Pro**virus **Act**ivity Detector âœ¨
æ ¹æ®å®¿ä¸»æµ‹åºåŸå§‹æ•°æ®è¯„ä¼°å…¶ä¸­åŸå™¬èŒä½“çš„æ´»è·ƒåº¦ã€‚

## ğŸ’  ç›®å½•

- [ProActæ£€æµ‹åŸç†](#-ProActæ£€æµ‹åŸç†)
- [ProActå·¥ä½œæµç¨‹](#-ProActå·¥ä½œæµç¨‹)
- [ä¾èµ–ç¯å¢ƒ](#-ä¾èµ–ç¯å¢ƒ)
- [å®‰è£…ä¸ç¯å¢ƒå‡†å¤‡](#-å®‰è£…ä¸ç¯å¢ƒå‡†å¤‡)
- [ä½¿ç”¨è¯´æ˜](#-ä½¿ç”¨è¯´æ˜)
  - [è¾“å…¥å‚æ•°](#-è¾“å…¥å‚æ•°)
  - [å‘½ä»¤ç¤ºä¾‹](#-å‘½ä»¤ç¤ºä¾‹)
  - [è¾“å‡ºæ–‡ä»¶](#-è¾“å‡ºæ–‡ä»¶)
- [æµ‹è¯•æ•°æ®](#-æµ‹è¯•æ•°æ®)
- [Citation](#-citation)
- [Contact](#-contact)

---

# Introduction

## ğŸ’  Principle
Using whole-genome sequencing (WGS) data, ProAct exploits the principle that a provirus in lysogeny shares the same copy number as its host, resulting in a Provirus-to-Host coverage ratio (PtoH) of 1, whereas transition toward lysis drives self-replication and elevates PtoH above 1.

<img width="803" height="197" alt="schematic" src="https://github.com/user-attachments/assets/05ddcefd-5bdb-4298-8e27-0fe7ee55f065" />

## ğŸ’  Workflow of ProBord
`ProAct` éœ€è¦è¾“å…¥å®¿ä¸»å‚è€ƒåŸºå› ç»„åŠå…¶åŸå§‹æµ‹åºæ•°æ®ã€åŸå™¬èŒä½“èµ·å§‹ä½ç‚¹ï¼Œé€šè¿‡ï¼ˆ1ï¼‰è´¨æ§è¿‡æ»¤åçš„åŸå§‹è¯»æ®µå’Œå‚è€ƒåŸºå› ç»„è¿›è¡Œæ¯”å¯¹ï¼Œç”Ÿæˆè¦†ç›–æ·±åº¦æ•°æ®ï¼›ï¼ˆ2ï¼‰åˆ†åˆ«è®¡ç®—æ¯ä¸ªmarker geneåŒºåŸŸçš„å¹³å‡coverageï¼Œå–ä¸­ä½å€¼ä»£è¡¨å®¿ä¸»coverageï¼›è®¡ç®—åŸå™¬èŒä½“åŒºåŸŸçš„å¹³å‡coverageä»£è¡¨å™¬èŒä½“coverageï¼›ï¼ˆ3ï¼‰è®¡ç®—PtoHï¼Œå¾—åˆ°è¯¥å®¿ä¸»å†…åŸå™¬èŒä½“çš„æ´»è·ƒåº¦ï¼ˆé€šè¿‡PtoHè¡¨å¾ï¼‰ã€‚

<img width="787" height="199" alt="workflow" src="https://github.com/user-attachments/assets/6c22bc29-d1eb-40cd-ad99-8127762a3adf" />

# Instructions

## ğŸ’  Dependencies
```
python3
bbmap
bwa
samtools
hmmer
prodigal
pandas
biopython
pysam
```

## ğŸ’  Installation
- Install miniconda and add channels (If already installed, please skip)
```
wget -c https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
bash Miniconda3-latest-Linux-x86_64.sh
source ~/.bashrc
conda config --add channels bioconda
conda config --add channels conda-forge
```
- Install dependencies
``` 
conda create -n proact python3 bbmap bwa samtools hmmer prodigal 
conda activate probord
pip install pandas biopython pysam
```
- Download ProAct from github
```
git clone https://github.com/mujiezhang/ProAct.git
```

## ğŸ’  How to run

- â–¶ï¸ Command line options
```
usage: proact_pipeline.py [-h] -g GENOME -1 READ1 -2 READ2 -p PHAGE_INFO [-o OUTPUT_DIR] [-t THREADS]
                          [-m] [--keep-tmp]

ProAct: Provirus Activity Detector

options:
  -h, --help            show this help message and exit
  -g, --genome GENOME   Host genome FASTA file
  -1, --read1 READ1     Read1 FASTQ file
  -2, --read2 READ2     Read2 FASTQ file
  -p, --phage-info PHAGE_INFO
                        A tab-delimited file with columns: virla_name, host_contig, start, end
  -o, --output-dir OUTPUT_DIR
                        Output directory (default: proact_results)
  -t, --threads THREADS
                        Number of threads (default: 40)
  -m, --meta            Run prodigal in meta mode (for metagenomes)
  --keep-tmp            Keep temporary files

Examples:
  # Basic usage
  python proact_pipeline.py -g host_genome.fna -1 sample_R1.fastq -2 sample_R2.fastq -p phage_info.tsv

  # Specify output dir, threads and meta mode
  python proact_pipeline.py  -g genome.fna  -1 reads_R1.fq  -2 reads_R2.fq -p phage_regions.tsv -o results_sample1 -t 40 -m --keep-tmp

Outputs:
  marker_genes.tsv              - GTDB marker genes list
  marker_gene_counts.tsv        - depth stats for marker genes
  phage_counts.tsv              - depth stats for phage regions
  host_counts.tsv               - median depth of marker genes
  PtoH_results.tsv              - final prediction results
```

Download test data:

```
wget -c https://zenodo.org/records/16948956/files/test-data.tar.gz
tar -xzvf test-data.tar.gz
```

Files in test-data:

```
test-data
â”œâ”€â”€ ERR4552622_R1.fastq 
â”œâ”€â”€ ERR4552622_R2.fastq
â”œâ”€â”€ GCA_904129595.1_S.Tm_LT2p22_assembled_genomic.fna
â””â”€â”€ phage_info.tsv
```
**Note**: `ERR4552622_R1.fastq` and `ERR4552622_R2.fastq` are the raw reads; `GCA_904129595.1_S.Tm_LT2p22_assembled_genomic.fna` is the host genome; `phage_info.tsv` contains location information for `phage p22`, `phage Fels-1` and `phage Fels-2`. We get these information from paper `Turkington C J R, Abadi N N, Edwards R A, et al. hafeZ: active prophage identification through read mapping[J]. bioRxiv, 2021: 2021.07. 21.453177.`

- run an example
```
python ProAct/proact_pipeline.py -g test-data/GCA_904129595.1_S.Tm_LT2p22_assembled_genomic.fna -1 test-data/ERR4552622_R1.fastq -2 test-data/ERR4552622_R2.fastq -p test-data/phage_info.tsv -t 40 -o test-result
```

## Output files
In this example, the results of ProBord's analysis will be written to the `test-result` directory, which will look like this:
```
test-result/
â”œâ”€â”€ host_counts.tsv
â”œâ”€â”€ marker_gene_counts.tsv
â”œâ”€â”€ marker_genes.tsv
â”œâ”€â”€ phage_counts.tsv
â””â”€â”€ PtoH_results.tsv
```
1. `host_counts.tsv`: median depth of host marker gene
2. `marker_gene_counts.tsv`: depth information of host marker gene
3. `marker_genes.tsv`: identified host marker gene
4. `phage_counts.tsv`: depth information of host marker gene
5. `PtoH_results.tsv`: prediction result

A detailed overview of `PtoH_results.tsv`:

| Host | Phage_Id | Contig | Start | Stop | Total_Counts | Ave_Counts | Median_of_MG | PtoH | Reads_depth_quality |
|------|----------|--------|-------|------|--------------|------------|--------------|------|---------------------|
| GCA_904129595.1_S.Tm_LT2p22_assembled_genomic | Sal_p22 | LR881463.1 | 1213987 | 1255756 | 149069351.0 | 3568.8137658606656 | 217.32838283828383 | 16.421296285613348 | high |
| GCA_904129595.1_S.Tm_LT2p22_assembled_genomic | Sal_Fels-1 | LR881463.1 | 1849458 | 1892188 | 39795679.0 | 931.3069902412768 | 217.32838283828383 | 4.285252474060286 | high |
| GCA_904129595.1_S.Tm_LT2p22_assembled_genomic | Sal_Fels-2 | LR881463.1 | 3731215 | 3764954 | 9870485.0 | 292.54549496147007 | 217.32838283828383 | 1.3460988902639377 | high |

- **`sample.depth`**ï¼šå‚è€ƒåŸºå› ç»„ä¸Šæ¯ä¸ªä½ç‚¹çš„æµ‹åºæ·±åº¦ï¼ˆchromosome, position, depthï¼‰ã€‚
- **`sample_phage_info.txt`**ï¼šæ ¹æ®ç”¨æˆ·è¾“å…¥ä¿¡æ¯æ±‡æ€»ï¼ˆsample, genome, contig, start-end, phage_idï¼‰ã€‚
- **`MG/`**ï¼šGTDB-Tk æ ‡æ³¨ç»“æœç›®å½•ï¼ŒåŒ…å«ï¼š
  - `genome.tsv`ï¼šmarker gene æ³¨é‡Šè¡¨
  - `gtdbtk.json`ï¼šæ³¨é‡Šå‚æ•°ä¸ç‰ˆæœ¬ä¿¡æ¯
  - `gtdbtk.log` / `gtdbtk.warnings.log`ï¼šè¿è¡Œæ—¥å¿—ä¸è­¦å‘Š
  - `identify/â€¦`ï¼šHMM æ¯”å¯¹ä¸ä¸­é—´æ–‡ä»¶
- **`counts/`**ï¼šæ·±åº¦ç»Ÿè®¡ç»“æœç›®å½•ï¼ŒåŒ…å«ï¼š
  - `marker_gene_counts.tsv`ï¼šå®¿ä¸»åŸºå› ç»„ä¸­æ¯ä¸ª marker gene çš„ä¿¡æ¯ï¼ˆGene Id, Total_Counts, Per_Counts, Median_Depth, Region_Lengthï¼‰
  - `phage_counts.tsv`ï¼šæ¯ä¸ªå™¬èŒä½“çš„ä¿¡æ¯ï¼ˆPhage_Id, Chromosome, Start, Stop, Total_Counts, Per_Counts, Median_Depth, Region_Lengthï¼‰
  - `host_counts.tsv`ï¼šå®¿ä¸»åŸºå› ç»„ä¸­æ‰€æœ‰ marker gene å¹³å‡æ·±åº¦çš„ä¸­ä½å€¼ï¼ˆSample_ID, Median_of_MGï¼‰
- **`PtoH.tsv`**ï¼šæœ€ç»ˆè¾“å‡ºï¼Œå°† `phage_counts.tsv`çš„`Per_Counts` é™¤ä»¥ `host_counts.tsv`çš„`Median_of_MG` å¾—åˆ° PtoH å€¼ï¼Œå¹¶é™„åŠ è´¨é‡æ ‡ç­¾ï¼ˆ`high`/`low`ï¼‰ã€‚å¦‚æœphage_Per_Counts > 10 ä¸” host_Median_of_MG > 10ï¼Œåˆ™ä¸º'high'ï¼›å¦åˆ™ä¸º'low'ï¼›å¦‚æœPtoH â‰¥ 1.5ï¼Œåˆ™æ´»æ€§åˆ¤å®šä¸º"active"ï¼Œè‹¥PtoH < 1.5ï¼Œåˆ™æ´»æ€§åˆ¤å®šä¸º"inactive".

  ğŸ’¡ **Note 2:** 
  å¦‚æœphage_Per_Counts < 10 æˆ– host_Median_of_MG < 10ï¼Œåˆ™è´¨é‡æ ‡ç­¾ä¸º'low'ã€‚è¡¨æ˜æµ‹åºæ·±åº¦ä¸è¶³ï¼Œå®¹æ˜“å¯¼è‡´PtoHåå·®è¾ƒå¤§ã€‚

# ğŸ’  Citation
......

# ğŸ“¬ Contact
```
# Mujie Zhang
# School of Life Sciences & Biotechnology, Shanghai Jiao Tong University
# Email: zhangmujie@sjtu.edu.cn
```
